src/compositions/cascade/useCascadePanels.ts

```diff
@@ -0,0 +1,178 @@
+import { computed, ComputedRef, inject, InjectionKey, MaybeRefOrGetter, provide, reactive, readonly, Ref, ref, toRef } from 'vue'
+import { isDefined } from '@/utilities'
+
+export const cascadePanelsKey: InjectionKey<UseCascadePanels> = Symbol('UseCascadePanels')
+
+export type CascadePanelId = string | symbol | number
+export type CascadeState = Record<CascadePanelId, boolean>
+
+export type CascadePanel = {
+  id: CascadePanelId,
+  level?: number,
+}
+
+export function getInjectedCascadePanels(): UseCascadePanels {
+  const cascadePanels = inject(cascadePanelsKey)
+
+  if (!cascadePanels) {
+    throw new Error('Cascade panels not found. Are you sure the component calling useCascadePanels() exists within a <p-cascade> component context?')
+  }
+
+  return cascadePanels
+}
+
+export type UseCascadePanels = {
+  openPanels: ComputedRef<CascadePanel[]>,
+  state: Readonly<CascadeState>,
+  isOpen: Ref<boolean>,
+  panelIsOpen: ComputedRef<(id: CascadePanelId) => boolean>,
+  getPanelById: (id: CascadePanelId) => CascadePanel | undefined,
+  openPanelById: (id: CascadePanelId) => void,
+  closePanelById: (id: CascadePanelId) => void,
+  togglePanelById: (id: CascadePanelId) => void,
+  closeAll: () => void,
+  close: () => void,
+  open: () => void,
+  toggle: () => void,
+}
+
+/**
+ * Provides a mechanism to manage a cascade of panels, each identified by a unique id, with operations to open, close, and toggle their states.
+ *
+ * @param {CascadePanel[]} panelsRefOrGetter - Initial array of panel definitions.
+ * @returns {UseCascadePanels} - An object including:
+ *   - `openPanels`: Computed array of open panels.
+ *   - `state`: Readonly object tracking the open/close state of each panel by id.
+ *   - `isOpen`: Ref boolean indicating whether the cascade panel group is open.
+ *   - `panelIsOpen`: Computed function to check if a panel is open by id.
+ *   - `getPanelById`: Function to get a panel by id.
+ *   - `openPanelById`: Function to open a panel by id.
+ *   - `closePanelById`: Function to close a panel by id.
+ *   - `togglePanelById`: Function to toggle the open state of a panel by id.
+ *   - `closeAll`: Function to close all panels.
+ *   - `close`, `open`, `toggle`: Functions to control the overall cascade state.
+ */
+export function useCascadePanels(panelsRefOrGetter?: MaybeRefOrGetter<CascadePanel[]>): UseCascadePanels {
+  if (!panelsRefOrGetter) {
+    try {
+      return getInjectedCascadePanels()
+    } catch {
+      throw new Error('No panels were provided; and no cascade panel context was found. Are you sure the component calling useCascadePanels() exists within a <p-cascade> component context?')
+    }
+  }
+
+  const panels = toRef(panelsRefOrGetter)
+  const state = reactive<CascadeState>(Object.fromEntries(panels.value.map((panel) => [panel.id, false])))
```

Comment:
`panelsRefOrGetter` is reactive data but its being used in a non reactive way here. If that value changes this state object won't update to reflect any panels that were added or removed.

Reply:
Ah we don't need to be doing this at all - updated to reflect that. The state is really only important for tracking if panels are open. If panels exist but have never been opened, they don't need to be in the `state` object yet but this made it look like they did